plumbing/http/http.go

```diff
@@ -0,0 +1,327 @@
+			if err != nil {
+				logf(h.ErrorLog, "error creating endpoint: %v", err)
+				renderStatusError(w, http.StatusBadRequest)
+				return
+			}
+
+			st, err := h.Loader.Load(ep)
+			if err != nil {
+				logf(h.ErrorLog, "error loading repository: %v", err)
+				renderStatusError(w, http.StatusNotFound)
+				return
+			}
+
+			ctx := r.Context()
+			ctx = context.WithValue(ctx, contextKey("errorLog"), h.ErrorLog)
+			ctx = context.WithValue(ctx, contextKey("repo"), m[1])
+			ctx = context.WithValue(ctx, contextKey("file"), file)
+			ctx = context.WithValue(ctx, contextKey("service"), service.svc)
+			ctx = context.WithValue(ctx, contextKey("storer"), st)
+			ctx = context.WithValue(ctx, contextKey("endpoint"), ep)
+
+			service.handler(w, r.WithContext(ctx))
+		}
+	}
+}
+
+func serviceRpc(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	st := ctx.Value(contextKey("storer")).(storage.Storer)
+	svc := ctx.Value(contextKey("service")).(transport.Service)
+	errorLog := ctx.Value(contextKey("errorLog")).(*log.Logger)
+	version := r.Header.Get("Git-Protocol")
+	contentType := r.Header.Get("Content-Type")
+
+	if contentType != fmt.Sprintf("application/x-git-%s-request", svc.Name()) {
+		renderStatusError(w, http.StatusForbidden)
+		return
+	}
+
+	w.Header().Set("Content-Type", fmt.Sprintf("application/x-git-%s-result", svc.Name()))
+	w.Header().Set("Connection", "Keep-Alive")
+	w.Header().Set("Transfer-Encoding", "chunked")
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+
+	var in bytes.Buffer
+	var reader io.ReadCloser
+	var err error
+	switch r.Header.Get("Content-Encoding") {
+	case "gzip":
+		reader, err = gzip.NewReader(r.Body)
+		if err != nil {
+			logf(errorLog, "error creating gzip reader: %v", err)
+			renderStatusError(w, http.StatusInternalServerError)
+			return
+		}
+		defer reader.Close() //nolint:errcheck
+	default:
+		reader = r.Body
+	}
+
+	io.Copy(&in, reader) //nolint:errcheck
+
+	var out bytes.Buffer
+	switch svc {
+	case transport.UploadPackService:
+		err = transport.UploadPack(ctx, st, io.NopCloser(&in), ioutil.WriteNopCloser(&out),
+			&transport.UploadPackOptions{
+				GitProtocol:   version,
+				AdvertiseRefs: false,
+				StatelessRPC:  true,
+			})
+	case transport.ReceivePackService:
+		err = transport.ReceivePack(ctx, st, io.NopCloser(&in), ioutil.WriteNopCloser(&out),
+			&transport.ReceivePackOptions{
+				GitProtocol:   version,
+				AdvertiseRefs: false,
+				StatelessRPC:  true,
+			})
+	}
+	if err != nil {
+		logf(errorLog, "error processing request: %v", err)
+		renderStatusError(w, http.StatusInternalServerError)
+		return
+	}
+
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		logf(errorLog, "http writer must implement http.Flusher interface")
+		// The writer must implement the Flusher interface which the default
+		// http.ResponseWriter does.
+		renderStatusError(w, http.StatusBadRequest)
+		return
+	}
+
+	// Write the response to the client in chunks.
+	p := make([]byte, 1024)
```

Comment:
Should we still chunk it when we can return the sideband error?
My answer is no, we should answer from the start

Reply:
ok. Can we maybe do it concurently? (Just a question)

Reply:
I'm not sure, honestly. HTTP1 only handles one stream at a time, but using HTTP2 _may_ benefit from this since it can handle multiple streams at once.